data/transactions/verify/txn.go

```diff
@@ -314,3 +314,113 @@ func LogicSig(txn *transactions.SignedTxn, ctx *Context) error {
 	return nil
 
 }
+
+// PaysetGroups verifies that the payset have a good signature and that the underlying
+// transactions are properly constructed.
+// Note that this does not check whether a payset is valid against the ledger:
+// a PaysetGroups may be well-formed, but a payset might contain an overspend.
+//
+// This version of verify is performing the verification over the provided execution pool.
+func PaysetGroups(ctx context.Context, payset [][]transactions.SignedTxn, blk bookkeeping.Block, verificationPool execpool.BacklogPool) (err error) {
+	txnGroupIdx := 0
+	const txnPerGroupThreshold = 32
+	proto, ok := config.Consensus[blk.BlockHeader.CurrentProtocol]
+	if !ok {
+		return protocol.Error(blk.BlockHeader.CurrentProtocol)
+	}
+	if len(payset) == 0 {
+		return nil
+	}
+	spec := transactions.SpecialAddresses{
+		FeeSink:     blk.BlockHeader.FeeSink,
+		RewardsPool: blk.BlockHeader.RewardsPool,
+	}
+	var txnGroups [][]transactions.SignedTxn
+	// prepare up to 16 concurrent worksets.
+	worksets := make(chan struct{}, 16)
+	worksDoneCh := make(chan interface{}, 16)
+	zeroAddress := basics.Address{}
+	processing := 0
+	tasksCtx, cancelTasksCtx := context.WithCancel(ctx)
+	defer cancelTasksCtx()
+	for processing >= 0 {
```

Comment:
It's clever that producer and consumer are both in this loop, but I think it would have been clearer split out. Probably spawn a go routine for the producer and then run the consumer loop in this function below that.

Reply:
I don't think that making the producer a separate go-routine makes sense here. The time we're spending on the producer is by order of several magnitudes smaller than the consumer. i.e. the loop is waiting in the select statement almost fo the entire time.